File: libs/engine/src/cache/cache.cc
#include <hestia/engine/cache.h>

#include <atomic>
#include <cctype>
#include <string>
#include <system_error>
#include <utility>

#include <spdlog/spdlog.h>

namespace hestia::engine {
    namespace fs = std::filesystem;

    namespace {
        std::string lower_hex(const std::string &hex) {
            std::string out = hex;
            for (auto &c: out) c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
            return out;
        }

        fs::path blob_path(const fs::path &dir, const proto::Checksum &checksum) {
            const std::string hex = lower_hex(checksum.hex);
            return dir / proto::to_string(checksum.algorithm) / hex.substr(0, 2) / hex;
        }

        void remove_quietly(const fs::path &path) {
            std::error_code ec;
            fs::remove(path, ec);
        }
    } // namespace

    Cache::Cache(fs::path dir) : dir_(std::move(dir)) {}

    fs::path Cache::dir() const {
        std::scoped_lock const lk(mu_);
        return dir_;
    }

    void Cache::reload(fs::path dir) {
        std::scoped_lock const lk(mu_);
        dir_ = std::move(dir);
    }

    std::optional<fs::path> Cache::lookup(const proto::Checksum &checksum) const {
        if (!proto::is_valid_checksum(checksum)) return std::nullopt;
        fs::path blob = blob_path(dir(), checksum);
        std::error_code ec;
        if (!fs::is_regular_file(blob, ec)) return std::nullopt;
        return blob;
    }

    void Cache::store(const fs::path &file, const proto::Checksum &checksum) {
        if (!proto::is_valid_checksum(checksum)) return;
        const fs::path blob = blob_path(dir(), checksum);
        std::error_code ec;
        if (fs::exists(blob, ec)) return;
        fs::create_directories(blob.parent_path(), ec);
        if (ec) return;

        static std::atomic<int> counter{0};
        const fs::path tmp = blob.string() + ".part" + std::to_string(++counter);
        fs::copy_file(file, tmp, fs::copy_options::overwrite_existing, ec);
        if (ec) {
            remove_quietly(tmp);
            return;
        }
        fs::rename(tmp, blob, ec);
        if (ec) {
            remove_quietly(tmp);
            return;
        }
        spdlog::debug("cached {} ({})", checksum.hex, proto::to_string(checksum.algorithm));
    }

    void Cache::evict(const proto::Checksum &checksum) {
        if (!proto::is_valid_checksum(checksum)) return;
        spdlog::debug("evicting cache blob {}", checksum.hex);
        remove_quietly(blob_path(dir(), checksum));
    }

    std::vector<CacheEntry> Cache::entries() const {
        std::vector<CacheEntry> out;
        const fs::path base = dir();
        std::error_code ec;
        for (const auto algorithm: {proto::HashAlgorithm::sha1, proto::HashAlgorithm::sha256}) {
            const fs::path root = base / proto::to_string(algorithm);
            for (fs::recursive_directory_iterator it(root, ec), end; !ec && it != end; it.increment(ec)) {
                if (!it->is_regular_file(ec)) continue;
                const proto::Checksum checksum{.algorithm = algorithm, .hex = it->path().filename().string()};
                if (!proto::is_valid_checksum(checksum)) continue;
                std::error_code size_ec;
                const auto size = it->file_size(size_ec);
                if (size_ec) continue;
                out.push_back(CacheEntry{.checksum = checksum, .size = size});
            }
            ec.clear();
        }
        return out;
    }

    CacheUsage Cache::usage() const {
        CacheUsage usage;
        for (const auto &entry: entries()) {
            ++usage.entries;
            usage.bytes += entry.size;
        }
        return usage;
    }

    CacheUsage Cache::clear() {
        const CacheUsage freed = usage();
        std::error_code ec;
        for (const auto algorithm: {proto::HashAlgorithm::sha1, proto::HashAlgorithm::sha256}) {
            fs::remove_all(dir() / proto::to_string(algorithm), ec);
        }
        spdlog::info("cleared cache: {} entries, {} bytes freed", freed.entries, freed.bytes);
        return freed;
    }
} // namespace hestia::engine

File: crates/engine/src/cache.rs
//! Content-addressed store of verified downloads, keyed by checksum
//! (`<dir>/<algorithm>/<hex[0:2]>/<hex>`). Blobs are immutable; consumers
//! re-verify on the way out, so a damaged blob is evicted, never served.

use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Mutex;

use proto::download::{Checksum, HashAlgorithm};

pub struct CacheEntry {
    pub checksum: Checksum,
    pub size: u64,
}

#[derive(Default, Clone, Copy)]
pub struct CacheUsage {
    pub entries: u64,
    pub bytes: u64,
}

pub struct Cache {
    dir: Mutex<PathBuf>,
}

const ALGORITHMS: [HashAlgorithm; 2] = [HashAlgorithm::Sha1, HashAlgorithm::Sha256];

impl Cache {
    pub fn new(dir: PathBuf) -> Self {
        Cache {
            dir: Mutex::new(dir),
        }
    }

    pub fn dir(&self) -> PathBuf {
        self.dir.lock().unwrap().clone()
    }

    pub fn reload(&self, dir: PathBuf) {
        *self.dir.lock().unwrap() = dir;
    }

    pub fn lookup(&self, checksum: &Checksum) -> Option<PathBuf> {
        if !checksum.is_valid() {
            return None;
        }
        let blob = blob_path(&self.dir(), checksum);
        blob.is_file().then_some(blob)
    }

    /// Best effort: a failure to cache never fails the download that fed it.
    pub fn store(&self, file: &Path, checksum: &Checksum) {
        if !checksum.is_valid() {
            return;
        }
        let blob = blob_path(&self.dir(), checksum);
        if blob.exists() {
            return;
        }
        let Some(parent) = blob.parent() else { return };
        if std::fs::create_dir_all(parent).is_err() {
            return;
        }
        static COUNTER: AtomicU64 = AtomicU64::new(0);
        let n = COUNTER.fetch_add(1, Ordering::Relaxed) + 1;
        let tmp = blob.with_extension(format!("part{n}"));
        if let Err(e) = std::fs::copy(file, &tmp) {
            tracing::warn!(hex = %checksum.hex, "failed to stage cache blob: {e}");
            let _ = std::fs::remove_file(&tmp);
            return;
        }
        if std::fs::rename(&tmp, &blob).is_err() {
            let _ = std::fs::remove_file(&tmp);
            return;
        }
        tracing::debug!(algorithm = checksum.algorithm.as_str(), hex = %checksum.hex, "cached blob");
    }

    pub fn evict(&self, checksum: &Checksum) {
        if !checksum.is_valid() {
            return;
        }
        let _ = std::fs::remove_file(blob_path(&self.dir(), checksum));
    }

    pub fn entries(&self) -> Vec<CacheEntry> {
        let base = self.dir();
        let mut out = Vec::new();
        for algorithm in ALGORITHMS {
            let root = base.join(algorithm.as_str());
            collect_entries(&root, algorithm, &mut out);
        }
        out
    }

    pub fn usage(&self) -> CacheUsage {
        let mut usage = CacheUsage::default();
        for entry in self.entries() {
            usage.entries += 1;
            usage.bytes += entry.size;
        }
        usage
    }

    pub fn clear(&self) -> CacheUsage {
        let freed = self.usage();
        let base = self.dir();
        for algorithm in ALGORITHMS {
            let _ = std::fs::remove_dir_all(base.join(algorithm.as_str()));
        }
        tracing::info!(
            entries = freed.entries,
            bytes = freed.bytes,
            "cache cleared"
        );
        freed
    }
}

fn blob_path(dir: &Path, checksum: &Checksum) -> PathBuf {
    let hex = checksum.hex.to_lowercase();
    dir.join(checksum.algorithm.as_str())
        .join(&hex[0..2])
        .join(&hex)
}

fn collect_entries(root: &Path, algorithm: HashAlgorithm, out: &mut Vec<CacheEntry>) {
    let Ok(shards) = std::fs::read_dir(root) else {
        return;
    };
    for shard in shards.flatten() {
        let Ok(blobs) = std::fs::read_dir(shard.path()) else {
            continue;
        };
        for blob in blobs.flatten() {
            let Ok(meta) = blob.metadata() else { continue };
            if !meta.is_file() {
                continue;
            }
            let hex = blob.file_name().to_string_lossy().into_owned();
            let checksum = Checksum { algorithm, hex };
            if checksum.is_valid() {
                out.push(CacheEntry {
                    checksum,
                    size: meta.len(),
                });
            }
        }
    }
}
